runtime/commonMain/src/kotlinx/serialization/ContextAware.kt

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2017-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.serialization
+
+import kotlinx.serialization.modules.*
+import kotlin.reflect.*
+
+/**
+ * Descriptor which allows retrieving additional runtime information, if proper [SerialModule] is provided.
+ *
+ * Usually, descriptors are generated by compiler plugin and know all information statically.
+ * However, that is not the case for [UnionKind.CONTEXTUAL] and [PolymorphicKind.OPEN] kinds of descriptors,
+ * where actual serializer used for property can be determined only at runtime.
+ * This descriptor is used by serializers which represent contextual serialization and open polymorphism (namely,
+ * [ContextSerializer] and [PolymorphicSerializer]), so they can capture statically known [kClass] in this descriptor.
+ * For schema introspection purposes, this [kClass] can be used in [SerialModule] to retrieve actual descriptor in runtime.
+ *
+ * This descriptor must always has either [UnionKind.CONTEXTUAL] or [PolymorphicKind.OPEN] kind,
+ * and vice versa: if some [SerialDescriptor.kind] is either CONTEXTUAL or OPEN, that descriptor
+ * could be safely casted to [ContextAwareDescriptor].
+ *
+ * ### Example
+ * Imagine we need to find all distinct properties names, which may occur in output after serializing a given class
+ * with respect to @[ContextualSerialization] annotation and all possible inheritors when the class is
+ * serialized polymorphically.
+ * Then we can write following function:
+ * ```
+ * fun allDistinctNames(descriptor: SerialDescriptor, module: SerialModule) = when (descriptor.kind) {
```

Comment:
Let's not advocate for using downcasting. 
`fun allDistinctNames(descriptor: ContextAwareDescriptor, module: SerialModule)` and let the authors decide what to do with non-`ContextAwareDescriptor`

Reply:
Then it would be impossible to generalize this function on a descriptors that are not of context kinds? e.g. `allDistinctNames(MyClass.serializer(), EmptyModule)`